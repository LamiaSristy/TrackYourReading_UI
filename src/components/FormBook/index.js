import React from 'react';
import PropTypes from 'prop-types';
import './FormBook.css';
import { connect } from 'react-redux';
import { updateBook } from '../../actions/book';

class FormBook extends React.Component {
  constructor(props) {
    super(props);
    this.state = {
      name: '',
      author: '',
      genre: '',
      pages: '',
    };
  }

  componentDidMount = () => {
    const { actionToPerform, buttonId, books } = this.props;
    if (actionToPerform === 'Save Changes') {
      const book = books.filter(x => x.id.toString() === buttonId);
      this.setState({
        name: book[0].name,
        author: book[0].author,
        genre: book[0].genre,
        pages: book[0].pages,
      });
    }
  }

  handleChangeName = e => {
    this.setState({
      name: e.target.value,
    });
  }

  handleChangeAuthor = e => {
    this.setState({
      author: e.target.value,
    });
  }

  handleChangeGenre = e => {
    this.setState({
      genre: e.target.value,
    });
  }

  handleChangePages = e => {
    this.setState({
      pages: e.target.value,
    });
  }

  handleSubmit = (name, author, genre, pages) => {
    const { addBook } = this.props;
    addBook(name, author, genre, pages);
  }

  handleUpdate = async id => {
    const { name, author, genre, pages } = this.state;

    const {
      user, updateBook, changeEditForm,
    } = this.props;

    const data = {
      id,
      user_id: user.user.id,
      name,
      author,
      genre,
      pages,
    };

    await updateBook(data);
    changeEditForm();
  }

  render() {
    const { name, author, genre, pages } = this.state;
    const {
      actionToPerform, books, buttonId, changeEditForm, changeAddForm,
    } = this.props;
    const book = books.filter(x => x.id.toString() === buttonId);
    return (
      <div>
        <h3>
          {actionToPerform}
          {' '}
          Book
        </h3>
        <form
          className="one-form"
          onSubmit={
           actionToPerform === 'Add'
             ? () => this.handleSubmit(name, author, genre, pages) : () => this.handleUpdate(book[0].id)
}
        >
          <div className="one-parameter">
            <label htmlFor="name">
              Name:
              <input
                required
                id="name"
                type="text"
                name="name"
                defaultValue={buttonId === '0' ? name : books[0].name}
                onChange={this.handleChangeName}
              />
            </label>
          </div>
          <div className="one-parameter">
            <label htmlFor="author">
              Author:
              <textarea
                id="author"
                name="author"
                defaultValue={buttonId === '0' ? author : books[0].author}
                onChange={this.handleChangeAuthor}
              />
            </label>
          </div>

          <div className="one-parameter">
            <label htmlFor="genre">
              genre
              <textarea
                id="genre"
                name="genre"
                defaultValue={buttonId === '0' ? genre : books[0].genre}
                onChange={this.handleChangeGenre}
              />
            </label>
          </div>
          <div className="one-parameter">
            <label htmlFor="pages">
              No of pages:
              <textarea
                id="pages"
                name="pages"
                defaultValue={buttonId === '0' ? pages : books[0].pages}
                onChange={this.handleChangePages}
              />
            </label>
          </div>
          <div className="buttons-form">
            {actionToPerform === 'Add' && <button type="submit">{actionToPerform}</button>}
            {actionToPerform === 'Save Changes' && <button type="submit">Save</button>}
            {actionToPerform === 'Add' && <button type="button" onClick={changeAddForm}>Cancel</button>}
            {actionToPerform === 'Save Changes' && <button type="button" onClick={changeEditForm}>Cancel</button>}
          </div>
        </form>
      </div>
    );
  }
}

FormBook.propTypes = {
  addBook: PropTypes.func,
  actionToPerform: PropTypes.string,
  changeAddForm: PropTypes.func,
  books: PropTypes.instanceOf(Array),
  buttonId: PropTypes.string,
  updateBook: PropTypes.func,
  changeEditForm: PropTypes.func,
  user: PropTypes.shape({
    user: PropTypes.shape({
      id: PropTypes.number,
    }),
  }),

};

FormBook.defaultProps = {
  changeAddForm: () => {},
  addBook: () => {},
  actionToPerform: '',
  books: [],
  buttonId: '0',
  updateBook: () => {},
  changeEditForm: () => {},
  user: {},
};

const mapStateToProps = state => ({
  user: state.user,
  books: state.books,
});
const mapDispatchToProps = dispatch => ({
  updateBook: data => dispatch(updateBook(data)),
});

export default connect(mapStateToProps, mapDispatchToProps)(FormBook);
