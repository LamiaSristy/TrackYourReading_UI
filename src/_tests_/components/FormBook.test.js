import React from 'react';
import { shallow } from 'enzyme';
import { Provider } from 'react-redux';
import configureMockStore from 'redux-mock-store';
import FormBook from '../../components/FormBook';

const mockStore = configureMockStore();
const store = mockStore({});
const setup = () => shallow(
  <Provider store={store}>
    <FormBook />
  </Provider>,
);

describe('FormBook Component', () => {
  let component;
  beforeEach(() => {
    component = setup();
  });

  test('renders the form the page', () => {
    const page = component.find('.one-form');
    expect(page).toMatchSnapshot();
  });

  test('renders the text of the one-parameter class', () => {
    const text = component.find('.one-parameter');
    expect(text).toMatchSnapshot();
  });

  test('renders the text of the buttons-form class', () => {
    const text = component.find('.buttons-form');
    expect(text).toMatchSnapshot();
  });
});
