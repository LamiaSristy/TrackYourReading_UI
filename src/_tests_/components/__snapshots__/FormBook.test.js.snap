// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`FormBook Component renders the form the page 1`] = `ShallowWrapper {}`;

exports[`FormBook Component renders the text of the buttons-form class 1`] = `ShallowWrapper {}`;

exports[`FormBook Component renders the text of the one-parameter class 1`] = `ShallowWrapper {}`;

exports[`Testpage Component renders the form the page 1`] = `ShallowWrapper {}`;

exports[`Testpage Component renders the text of the add-book class 1`] = `ShallowWrapper {}`;

exports[`Testpage Component renders the text of the book-description class 1`] = `ShallowWrapper {}`;

exports[`Testpage Component renders the text of the book-info class 1`] = `ShallowWrapper {}`;

exports[`Testpage Component renders the text of the book-name class 1`] = `ShallowWrapper {}`;

exports[`Testpage Component renders the text of the books class 1`] = `ShallowWrapper {}`;

exports[`Testpage Component renders the text of the buttons class 1`] = `ShallowWrapper {}`;

exports[`Testpage Component renders the text of the buttons-form class 1`] = `ShallowWrapper {}`;

exports[`Testpage Component renders the text of the one-book class 1`] = `ShallowWrapper {}`;

exports[`Testpage Component renders the text of the one-parameter class 1`] = `ShallowWrapper {}`;

exports[`Testpage Component renders the text of the tracking class 1`] = `ShallowWrapper {}`;

exports[`Testpage Component renders the the page 1`] = `ShallowWrapper {}`;
